### PR TITLE
Implement GetAgentAvailability Method

### PR DESCRIPTION
**Pull request by charliewang95**
Feature parity with GetAgentAvailability in LCW V1

image

### CODE REVIEW

File: src/OmnichannelChatSDK.ts [hunk 1]
```diff
@@ -1693,6 +1635,146 @@ class OmnichannelChatSDK {
         }
     }
 
+    public async getAgentAvailability(optionalParams: GetAgantAvailabilityOptionalParams = {}): Promise<any> { // eslint-disable-line @typescript-eslint/no-explicit-any
+        const requestId = uuidv4();
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:1639]: `requestId` is already generated on instantiation of of the `OmnichannelChatSDK` class. You can simple reuse `OmnichannelChatSDK.requestId`

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 2]
```diff
@@ -3296,5 +3296,162 @@ describe('Omnichannel Chat SDK', () => {
...
+
+            try {
+                await chatSDK.getAgentAvailability();
+                throw("Should throw error.");
+            } catch (ex) {
+                expect(ex?.message).toEqual("GetAgentAvailability can only be called before a chat has started.")
+            }
+        });
+
+        it('ChatSDK.getAgentAvailability() should throw error if conversation already started', async () => {
```
Comment (elopezanaya) [inline __tests__/OmnichannelChatSDK.spec.ts:3410]: is the name correct?, is the same as LoC 3372 but different logic.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,7 @@ All notable changes to this project will be documented in this file.
 - Add ability to use `ChatSDK.createChatAdapter()` for `DirectLine` protocol
 - Add `CreateACSAdapter` telemetry event
 - Improve `ChatSDK.createChatAdapter()` with retries using exponential backoff & additional details on failures
+- Add `GetAgentAvailability` SDK method
```
Comment (elopezanaya) [inline CHANGELOG.md:9]: my suggestion is to add that the method is exclusively for auth chats

File: src/OmnichannelChatSDK.ts [hunk 4]
```diff
@@ -1693,6 +1635,146 @@ class OmnichannelChatSDK {
...
+                message: "GetAgentAvailability can only be called before a chat has started."
+            }
+
+            this.scenarioMarker.failScenario(TelemetryEvent.GetAgentAvailability, {
+                RequestId: requestId,
+                ChatId: this.chatToken.chatId as string,
+                ExceptionDetails: JSON.stringify(exceptionDetails)
+            });
+
+            throw new Error(exceptionDetails.message);
```
Comment (elopezanaya) [inline src/OmnichannelChatSDK.ts:1685]: this code is repeated 3 times in this function, could we find a way to abstract it?

File: src/core/GetAgantAvailabilityOptionalParams.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import InitContext from "@microsoft/ocsdk/lib/Model/InitContext";
+
+export default interface GetAgantAvailabilityOptionalParams {
```
Comment (elopezanaya) [inline src/core/GetAgantAvailabilityOptionalParams.ts:3]: typo on Agent...

File: src/OmnichannelChatSDK.ts [hunk 6]
```diff
@@ -1693,6 +1635,146 @@ class OmnichannelChatSDK {
...
+        });
+
+        if (!this.authSettings) {
+            const exceptionDetails: ChatSDKExceptionDetails = {
+                response: "UnAuthUnsupported",
+                message: "GetAgentAvailability is supported only for authenticated live chat widget."
+            };
+
+            this.scenarioMarker.failScenario(TelemetryEvent.GetAgentAvailability, {
+                RequestId: requestId,
```
Comment (akagar-microsoft) [inline src/OmnichannelChatSDK.ts:1652]: Can we add Chat Id here ?

File: src/OmnichannelChatSDK.ts [hunk 7]
```diff
@@ -1693,6 +1635,146 @@ class OmnichannelChatSDK {
...
+        }
+
+        if (!this.authenticatedUserToken) {
+            const exceptionDetails = {
+                response: "UndefinedAuthToken",
+                message: "Missing AuthToken for GetAgentAvailability."
+            }
+
+            this.scenarioMarker.failScenario(TelemetryEvent.GetAgentAvailability, {
+                RequestId: requestId,
```
Comment (akagar-microsoft) [inline src/OmnichannelChatSDK.ts:1666]: Can we add chat id here ?

File: src/OmnichannelChatSDK.ts (hunk 1 above)
Comment (charliewang95) [reply src/OmnichannelChatSDK.ts:1639]: updated

File: __tests__/OmnichannelChatSDK.spec.ts (hunk 2 above)
Comment (charliewang95) [reply __tests__/OmnichannelChatSDK.spec.ts:3410]: fixed

File: CHANGELOG.md (hunk 3 above)
Comment (charliewang95) [reply CHANGELOG.md:9]: updated

File: src/OmnichannelChatSDK.ts (hunk 4 above)
Comment (charliewang95) [reply src/OmnichannelChatSDK.ts:1685]: refactored

File: src/OmnichannelChatSDK.ts (hunk 6 above)
Comment (charliewang95) [reply src/OmnichannelChatSDK.ts:1652]: This API is called before startChat, so no Chat Id is available yet

File: src/OmnichannelChatSDK.ts (hunk 7 above)
Comment (charliewang95) [reply src/OmnichannelChatSDK.ts:1666]: same as above

**Review by elopezanaya — APPROVED** [review]

File: src/OmnichannelChatSDK.ts [hunk 8]
```diff
@@ -1693,6 +1635,118 @@ class OmnichannelChatSDK {
...
+
+            throw new Error(exceptionDetails.message);
+        }
+
+        this.scenarioMarker.startScenario(TelemetryEvent.GetAgentAvailability, {
+            RequestId: this.requestId
+        });
+
+        if (!this.authSettings) {
+            reportError("UnAuthUnsupported", "GetAgentAvailability is supported only for authenticated live chat widget.");
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:1659]: Would `Unsupported` be better? We'll be able to reuse this type of response in other scenarios. We can be more descriptive in the message

**Pull request merged**